src/main/java/org/jabref/gui/citationkeypattern/GenerateCitationKeyAction.java

```diff
@@ -80,30 +92,54 @@ private void checkOverwriteKeysChosen() {
         }
     }
 
-    private void generateKeys() {
-        if (isCanceled) {
-            return;
-        }
-
-        stateManager.getActiveDatabase().ifPresent(databaseContext -> {
-            // generate the new citation keys for each entry
-            final NamedCompound compound = new NamedCompound(Localization.lang("Autogenerate citation keys"));
-            CitationKeyGenerator keyGenerator =
-                    new CitationKeyGenerator(databaseContext, Globals.prefs.getCitationKeyPatternPreferences());
-            for (BibEntry entry : entries) {
-                keyGenerator.generateAndSetKey(entry)
-                            .ifPresent(fieldChange -> compound.addEdit(new UndoableKeyChange(fieldChange)));
+    private BackgroundTask generateKeysInBackground() {
+        return new BackgroundTask<Void>() {
+
+            private NamedCompound compound;
+
+            @Override
+            protected Void call() {
+                    if (isCanceled) {
+                        return null;
+                    }
+                    DefaultTaskExecutor.runInJavaFXThread(() -> {
+                        updateProgress(0, entries.size());
+                        messageProperty().set(Localization.lang("%0/%1 entries", 0, entries.size()));
+                    });
+                    stateManager.getActiveDatabase().ifPresent(databaseContext -> {
+                        // generate the new citation keys for each entry
+                        compound = new NamedCompound(Localization.lang("Autogenerate citation keys"));
+                        CitationKeyGenerator keyGenerator =
+                                new CitationKeyGenerator(databaseContext, Globals.prefs.getCitationKeyPatternPreferences());
```

Comment:
Can you extract the call to Globals somehow?

Reply:
Should be possibel to pass it as constructor argument

Reply:
Yes!